File: libsrc/unistd/chown.c
/*
 * Copyright(C) 2011-2014 Pedro H. Penna <pedrohenriquepenna@gmail.com>
 * 
 * unistd/chown.c - chown() system call.
 */

#include <nanvix/syscall.h>
#include <unistd.h>
#include <errno.h>

/*
 * Changes owner and group of a file.
 */
int chown(const char *path, uid_t owner, gid_t group)
{
	int ret;
	
	__asm__ volatile (
		"int $0x80"
		: "=a" (ret)
		: "0" (NR_chown),
		  "b" (path),
		  "c" (owner),
		  "d" (group)
	);
	
	/* Error. */
	if (ret < 0)
	{
		errno = -ret;
		return (-1);
	}
	
	return (
	ret);
}

File: src/libs/posix/src/unistd/syscall/chown.rs
// Copyright(c) The Maintainers of Nanvix.
// Licensed under the MIT License.

//==================================================================================================
// Imports
//==================================================================================================

use crate::sys::types::{
    gid_t,
    uid_t,
};
use ::nvx::sys::error::Error;

//==================================================================================================
// Standalone Functions
//==================================================================================================

///
/// # Description
///
/// Changes the user and group ownership of a file.
///
/// # Parameters
///
/// - `path`: Path to the file.
/// - `owner`: User ID of the new owner.
/// - `group`: Group ID of the new owner.
///
/// # Returns
///
/// Upon successful completion, empty is returned. Otherwise, it returns an error.
///
pub fn chown(path: &str, owner: uid_t, group: gid_t) -> Result<(), Error> {
    ::nvx::trace!("chown(): path = {:?}, owner = {:?}, group = {:?}", path, owner, group);
    crate::unistd::fchownat(crate::fcntl::AT_FDCWD, path, owner, group, 0)
}
